### PR TITLE
Allow external recording using `quality` parameter

### LINKED ISSUE #4216: Can use external recorder app
Based on https://forum.getodk.org/t/form-spec-proposal-configure-audio-quality-for-internal-recording/31094

**Acceptance**

- [x] Given I'm using a form with an audio question with `quality` parameter as `foo`
When I click "Choose Sound"
Then it uses the in-app recording

- [x] Given I'm using a form with an audio question with `quality` parameter as `external`
When I click "Choose Sound"
Then it uses the default recorder app on my device

- [x] Given I've enabled "Use external recording app" in settings
When I record with an audio question
Then the recording is made using the default recorder app on my device

- [x] Given I've enabled "Use external recording app" in settings
When I record with an audio question that has `normal` or `voice-only` as `quality`
Then the recording is made using the in-app audio recorder

### PR DESCRIPTION
**Pull request by seadowg**
Closes #4216

#### What has been done to verify that this works as intended?

New tests and manually verified.

#### Why is this the best possible solution? Were any other approaches considered?

The implementation is actually pretty straight forward for this. I've left a couple of comments inline.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

I think we can skip QA for this until we have support for the other audio qualities in there.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/formentry/ODKView.java [hunk 1]
```diff
@@ -124,35 +127,27 @@
     PreferencesProvider preferencesProvider;
 
     @Inject
-    AudioRecorderViewModelFactory audioRecorderViewModelFactory;
+    ActivityAvailability activityAvailability;
+
+    private final WidgetFactory widgetFactory;
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formentry/ODKView.java:132]: It felt like it made sense to rearrange this relationship so the `ODKView` owns a `WidgetFactory` object. In my mind this makes the dependency clearer.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java [hunk 2]
```diff
@@ -124,4 +126,13 @@ public static String markQuestionIfIsRequired(String questionText, boolean isReq
 
         return questionText;
     }
+
+    public static Optional<String> getAttributeValue(FormEntryPrompt prompt, String attributeName) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: We talked briefly about using optionals vs sticking to nulls. Are we diving into optionals?

File: collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java [hunk 3]
```diff
@@ -194,11 +193,8 @@ public static QuestionWidget createWidgetFromPrompt(FormEntryPrompt prompt,
             case Constants.CONTROL_AUDIO_CAPTURE:
                 RecordingRequester recordingRequester;
 
-                Optional<TreeElement> audioQuality = prompt.getBindAttributes().stream().filter(attr -> {
-                    return attr.getName().equals("audio-quality");
-                }).findAny();
-
-                if (audioQuality.isPresent() && audioQuality.get().getAttributeValue().equals("external")) {
+                Optional<String> audioQuality = FormEntryPromptUtils.getAttributeValue(prompt, "audio-quality");
+                if (audioQuality.isPresent() && audioQuality.get().equals("external")) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/WidgetFactory.java]: This feels like a very `null` structure if you see what I mean. Not really much benefit to an `Optional` if you're not chaining, for example.

**Review by lognaturel — COMMENTED** [review]
One nitpicky question/conversation point about `Optional` since this would be the first one in the Java codebase.

I also realize we didn't talk about the upgrade experience and whether users who are already happily using external applications should get thrown into the internal recorder. I think it would be better for folks who are updating to continue with the existing behavior and new users to use internal recording by default. That seems like a separate story.

File: collect_app/src/main/java/org/odk/collect/android/widgets/utilities/RecordingRequesterFactory.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        this.audioRecorderViewModel = audioRecorderViewModel;
+        this.permissionUtils = permissionUtils;
+        this.activity = activity;
+        this.lifecycle = lifecycle;
+    }
+
+    public RecordingRequester create(FormEntryPrompt prompt, boolean externalRecorderPreferred) {
+        Optional<String> audioQuality = FormEntryPromptUtils.getAttributeValue(prompt, "quality");
+
+        if (audioQuality.isPresent() && (audioQuality.get().equals("normal") || audioQuality.get().equals("voice-only"))) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/utilities/RecordingRequesterFactory.java]: This feels like a very null structure if you see what I mean. Not really much benefit to an Optional if you're not chaining, for example.

File: collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java (hunk 2 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/utilities/FormEntryPromptUtils.java]: Ah yeah good point. I'm in two minds. I think your point below is a good one. Java Optionals really don't add much as calling `get` is super easy. I think I'm happy to convert the `Optional` from the stream into a normal type and using `Nullable`.

**Comment by seadowg** [comment]
> I think it would be better for folks who are updating to continue with the existing behavior and new users to use internal recording by default.

@lognaturel yeah I think I agree.  I'll create a story!

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
New upgrade story at https://github.com/getodk/collect/issues/4223!

**Pull request merged**